### ISSUE TITLE
SteamVR hand controllers ignored when using Oculus DK2

### ISSUE THREAD

**Issue by ReallyDiskun** [highfidelity/hifi#9068]
Hi folks,

I've come across an issue: when launching HiFi either from Desktop or Steam, it detects my Oculus DK2 as Oculus Rift and overrides SteamVR altogether. In fact, you can close SteamVR and HiFi will keep running nonchalantly. Could we have a switch or a flag to choose which VR API do we want to use and not hook the Oculus Runtime straight away? Or at least have it list all VR APIs instead of just Oculus.

This wouldn't bother me at all if I wasn't using [an open source hand controller driver called PSMoveService](https://github.com/cboulay/PSMoveService) that allows to use cheap Playstation Move controllers and PS3Eye cameras for hand control with SteamVR. This is pretty cool since it provides affordable access to moderately accurate hand control for pocket money. Also it's open source, which is cool and I have contributed to the project by making controller models and textures (C++ gives me headaches, sorry guys).

Here's me in Steam's The Lab with the DK2 and PS Move:
[http://i.imgur.com/tNPpgBD.gifv](http://i.imgur.com/tNPpgBD.gifv)

I've also tried Star Wars Trials on Tatooine, The Cubicle and others. As I said, totally compatible with SteamVR titles, but HiFi seems to be ignoring Steam when an Oculus Rift is plugged.

Benefits of using SteamVR instead of the Oculus Runtime:
- Grabbing stuff with my hands!
- Steam friends integration.
- Steam VR Home launcher is way better than Oculus Home.
- Taking snapshots with the motion controllers.

Already tried:
- Using a previous Oculus runtime (HMD not detected by SteamVR).
- Using no Oculus runtime (HiFi will not detect the HMD - duh).
- Using the "-vrmode OpenVR" commandline option available in other titles with HiFi (nothing).
- Various tinkering.

TL;DR Can we please have a way to force HiFi to use SteamVR or at least make it use the SteamVR controllers with an Oculus HMD?

**Comment by ReallyDiskun** [comment]
I've also tried to use the Revive wrapper to try and fool Interface into SteamVR. It crashed.

In the meanwhile I took a look at hifi-log.txt and this caught my attention:
`[11/14 23:37:34] [DEBUG] [hifi.controllers] Registered input device < "Hydra" > deviceID =  3
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 1 : "Keyboard"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 2 : "Neuron"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 3 : "Hydra"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 65280 : "Application"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 65281 : "Actions"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Registered input device < "OculusTouch" > deviceID =  4
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 1 : "Keyboard"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 2 : "Neuron"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 3 : "Hydra"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 4 : "OculusTouch"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 65280 : "Application"
[11/14 23:37:34] [DEBUG] [hifi.controllers] Device 65281 : "Actions"
[11/14 23:37:34] [DEBUG] [hifi.avatars] Changing skeleton model for avatar QUuid("{00000000-0000-0000-0000-000000000000}") to "http://diskungate.000webhostapp.com/Diskun.fst"
[11/14 23:37:34] [DEBUG] [hifi.networking] Starting request for: http://diskungate.000webhostapp.com/Diskun.fst
[11/14 23:37:34] [DEBUG] [hifi.avatars] Changing display name for avatar to ""
[11/14 23:37:34] [DEBUG] [hifi.interface] Faceshift: Disconnecting...
[11/14 23:37:34] [DEBUG] [hifi.interface] DDE Face Tracker: Socket:  "Bound"
[11/14 23:37:34] [WARNING] [default] QObject::connect: signal not found in Application
[11/14 23:37:34] [DEBUG] [hifi.interface] Startup time: 5.67 seconds.
[11/14 23:37:34] [DEBUG] [hifi.interface] Checking  "C:/Users/disku/AppData/Roaming/High Fidelity/assignment-client/content-version.txt"  for content version
[11/14 23:37:34] [DEBUG] [hifi.interface] Server content version:  22
**[11/14 23:37:34] [DEBUG] [hifi.interface] Has HMD + Hand Controllers:  false , current plugin:  "Oculus Rift"**
[11/14 23:37:34] [DEBUG] [hifi.interface] Has tutorial content:  true`

So there must be a way to change this, right? Please? 😃

**Comment by ZappoMan** [comment]
There's an undocumented, and not officially sorted environment variable that you can enable in your widows environment - that will we will look for to use the OpenVR/SteamVR sdk instead of oculus... you could try this, and it should do what you want.  `HIFI_DEBUG_OPENVR`

**Comment by ZappoMan** [comment]
I'd love to hear if that works for you.

**Comment by ReallyDiskun** [comment]
@ZappoMan Thank you very much. Correct me if I'm wrong. I guess that when you say env variable you mean I have to compile Interface setting that env var in Visual Studio? I'll try as soon as I can.

**Comment by ZappoMan** [comment]
@ReallyDiskun - you shouldn't need to recompile, the current official build should detect that environment variable and show the OpenVR option on the View menu.

**Comment by ZappoMan** [comment]
@ReallyDiskun - ok - scratch that idea - I just tested, and it's overly judder-- as I mentioned, it's not an officially supported feature. We'll take this feature request under advisement... maybe get back to the support for Oculus HMD inside of OpenVR or maybe some other solution for PSMove and controllers. We do have a plugin API for hand controllers, and so it may also be possible to simply support the move controllers natively.

**Comment by ReallyDiskun** [comment]
Indeed it worked, but with lots of judder and 100% CPU usage.

![image](https://cloud.githubusercontent.com/assets/20774380/20287895/b90545e4-aacf-11e6-9a38-5548ae5f90d9.png)

Disabling the preview didn't help. It's a shame, it would've been cool. Support for Oculus inside OpenVR would be cool. I do believe that the OpenVR way is much better than Oculus' proprietary stuff.

**Comment by ZappoMan** [comment]
@ReallyDiskun - let me get back with you on this.

**Comment by ZappoMan** [comment]
@ReallyDiskun - ok - I found a bug in our OpenVR display plugin code that is causing the crazy judder... it should be fixed in our next beta... but if you want to test it you can grab the installer off of this Pull Request... https://github.com/highfidelity/hifi/pull/9073

**Comment by ZappoMan** [comment]
It should be noted that this isn't an officially supported configuration. I do see a fair amount of judder while content is loading, so in the long run, this isn't the actual best approach. I'll let you be the judge of if it's useful for your needs or not.

**Comment by ReallyDiskun** [comment]
That was quick! Thank you once again @ZappoMan. I'll test the build once I'm back from work and let you know the result.

**Comment by ZappoMan** [comment]
The fix to the judder has been merged into master and will be available in the next beta.
